applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h

```diff
@@ -0,0 +1,435 @@
+     * Solve the generalized eigenvalue problem using FEAST
+     * @param rK first input matrix
+     * @param rM second input matrix
+     * @param rEigenvalues eigenvalues
+     * @param rEigenvectors row-aligned eigenvectors [n_evs,n_dofs]
+     */
+    void Solve(
+        SparseMatrixType& rK,
+        SparseMatrixType& rM,
+        VectorType& rEigenvalues,
+        DenseMatrixType& rEigenvectors) override
+    {
+        // settings
+        const size_t system_size = rK.size1();
+        size_t subspace_size;
+        const TScalarIn Ti = {};
+        const TScalarOut T = {};
+
+        if( mParam["search_lowest_eigenvalues"].GetBool() || mParam["search_highest_eigenvalues"].GetBool() )
+        {
+            subspace_size = 2 * static_cast<size_t>(mParam["number_of_eigenvalues"].GetInt());
+        }
+        else if( mParam["subspace_size"].GetInt() == 0 )
+        {
+            subspace_size = 1.5 * static_cast<size_t>(mParam["number_of_eigenvalues"].GetInt());
+        }
+        else
+        {
+            subspace_size = static_cast<size_t>(mParam["subspace_size"].GetInt());
+        }
+
+        // create column based matrix for the fortran routine
+        FEASTMatrixType tmp_eigenvectors(system_size, subspace_size);
+        OutputVectorType tmp_eigenvalues(subspace_size);
+        VectorType residual(subspace_size);
+
+        // set FEAST settings
+        int fpm[64] = {};
+        feastinit(fpm);
+
+        mParam["echo_level"].GetInt() > 0 ? fpm[0] = 1 : fpm[0] = 0;
+
+        fpm[2] = -std::log10(mParam["tolerance"].GetDouble());
+        fpm[3] = mParam["max_iteration"].GetInt();
+
+        // compute only right eigenvectors
+        if( !TSymmetric )
+            fpm[14] = 1;
+
+        if( mParam["search_lowest_eigenvalues"].GetBool() )
+        {
+            fpm[39] = -1;
+        }
+        if( mParam["search_highest_eigenvalues"].GetBool() )
+        {
+            fpm[39] = 1;
+        }
+
+        char UPLO = 'F';
+        int N = static_cast<int>(system_size);
+
+        // provide matrices in array form. fortran indices start with 1, must be int
+        double* A = reinterpret_cast<double*>(rK.value_data().begin());
+        std::vector<int> IA;
+        IA.reserve(N+1);
+        for( int i=0; i<N+1; ++i )
+        {
+            IA[i] = static_cast<int>(rK.index1_data()[i]) + 1;
+        }
+        std::vector<int> JA;
+        JA.reserve(IA[N]-1);
+        for( int i=0; i<IA[N]-1; ++i )
+        {
+            JA[i] = static_cast<int>(rK.index2_data()[i]) + 1;
+        }
+
+        double* B = reinterpret_cast<double*>(rM.value_data().begin());
+        std::vector<int> IB;
+        IB.reserve(N+1);
+        for( int i=0; i<N+1; ++i )
+        {
+            IB[i] = static_cast<int>(rM.index1_data()[i]) + 1;
+        }
+        std::vector<int> JB;
+        JB.reserve(IB[N]-1);
+        for( int i=0; i<IB[N]-1; ++i )
+        {
+            JB[i] = static_cast<int>(rM.index2_data()[i]) + 1;
+        }
```

Comment:
global question, in the lines above you seem to copy the matrices to the feast-needed input
Could you do sth like the `UblasWrapper` @oberbichler did for eigen such that this copying is not necessary?

Reply:
I am not sure if this is possible, as I have to cast from size_t to int, which do not have the same size. Just casting the array messes things up (I think we discussed about this specfic issue once)

Reply:
I don't think that you can use the wrapper because you also need one-based-indexing(?)

Reply:
plus the wrapper converts from ublas data to Eigen's Map classes, which is not what we need here